### PR TITLE
Feature toggleability

### PR DESCRIPTION
**Pull request by msohailhussain**
Datafile Parsing

Decision Service changes

Unit tests

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 1]
```diff
@@ -215,7 +215,7 @@ - (BOOL)isFeatureEnabled:(NSString *)featureKey userId:(NSString *)userId attrib
     
     OPTLYFeatureDecision *decision = [self.decisionService getVariationForFeature:featureFlag userId:userId attributes:attributes];
     
-    if (!decision) {
+    if (!decision || !decision.variation.featureEnabled) {
```
Comment (mikeproeng37) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:218]: Can we also null check the `variation` or does obj-c handle nil values gracefully?

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 2]
```diff
@@ -382,6 +382,72 @@ - (void)testIsFeatureEnabledWithFeatureFlagEnabledAndUserIsBeingExperimented {
     [decisionServiceMock stopMocking];
 }
 
+// Should return true if the feature experiment variation’s `featureEnabled` property is true, false otherwise
+- (void)testIsFeatureEnabledWithVariationsFeatureEnabledForFeatureFlag {
```
Comment (mikeproeng37) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: Can we split this up into two test cases

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 3]
```diff
@@ -382,6 +382,72 @@ - (void)testIsFeatureEnabledWithFeatureFlagEnabledAndUserIsBeingExperimented {
...
+    XCTAssertTrue([self.optimizely isFeatureEnabled:featureFlagKey userId:kUserId attributes:nil], @"should return true for enabled featureFlag");
+    
+    OCMVerify([decisionServiceMock getVariationForFeature:featureFlag userId:kUserId attributes:nil]);
+
+    [decisionServiceMock stopMocking];
+}
+
+// Should return true if the user is bucketed into rollout experiment’s variation
+// and variation's featureEnabled is also true, false otherwise
+- (void)testIsFeatureEnabledWithVariationsFeatureEnabledForRollout {
```
Comment (mikeproeng37) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: Same as above

**Review by mikeproeng37 — CHANGES_REQUESTED** [review]
Mostly looks good, I just have a few comments

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m (hunk 1 above)
Comment (msohailhussain) [reply OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:218]: we are checking null variation in the GetVariationForFeature method. I guess we shouldn't check here.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 4]
```diff
@@ -382,6 +382,84 @@ - (void)testIsFeatureEnabledWithFeatureFlagEnabledAndUserIsBeingExperimented {
     [decisionServiceMock stopMocking];
 }
 
+// Should return false if the feature experiment variation’s `featureEnabled` property is true
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: The code comment seems to have a little error.  Our design doc says:

Test that `isFeatureEnabled` returns false if the feature experiment variation’s `featureEnabled` property is false

The last word in the code comment should be "false", not "true".

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 5]
```diff
@@ -382,6 +382,84 @@ - (void)testIsFeatureEnabledWithFeatureFlagEnabledAndUserIsBeingExperimented {
...
+    
+    XCTAssertTrue([self.optimizely isFeatureEnabled:featureFlagKey userId:kUserId attributes:nil], @"should return true for enabled featureFlag");
+    
+    OCMVerify([decisionServiceMock getVariationForFeature:featureFlag userId:kUserId attributes:nil]);
+    
+    [decisionServiceMock stopMocking];
+}
+
+// Should return true if the user is bucketed into rollout experiment’s variation
+// and variation's featureEnabled is also true
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:423]: Comment on the code comment.  Our design doc requests:

Test that `isFeatureEnabled` returns true if the user is bucketed into the rollout experiment’s variation

I've also read in the design doc that "featureEnabled only appears in variations under experiments
which are feature experiments".  I infer from the design doc that rollout experiment variations will
not contain the key "featureEnabled".  It is the decision service response from `isFeatureEnabled`
which will be true.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m [hunk 6]
```diff
@@ -382,6 +382,84 @@ - (void)testIsFeatureEnabledWithFeatureFlagEnabledAndUserIsBeingExperimented {
...
+    
+    XCTAssertTrue([self.optimizely isFeatureEnabled:featureFlagKey userId:kUserId attributes:nil], @"should return true for enabled featureFlag");
+    
+    OCMVerify([decisionServiceMock getVariationForFeature:featureFlag userId:kUserId attributes:nil]);
+    [decisionServiceMock stopMocking];
+}
+
+// Should return false if the user is bucketed into rollout experiment’s variation
+// but variation's featureEnabled is false
+- (void)testIsFeatureEnabledWithVariationsFeatureEnabledForRollout {
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:444]: Comment on the code comment very nearly like the preceding comment on a code comment.  In this case:

Comment on the code comment.  Our design doc requests:

Test that `isFeatureEnabled` returns false if the user is NOT bucketed into the rollout experiment’s variation

I've also read in the design doc that "featureEnabled only appears in variations under experiments
which are feature experiments".  I infer from the design doc that rollout experiment variations will
not contain the key "featureEnabled".  It is the decision service response from `isFeatureEnabled`
which will be false.

Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m]: CAN UNIT TESTS SHOULD CHECK FOR PRESENCE/ABSENCE OF "featureEnabled"?

If the design doc's statement "featureEnabled only appears in variations under experiments
which are feature experiments" does mean "rollout experiment" variations don't contain
"featureEnabled", then I think it is best if we can arrange our unit tests to verify this claim
for both "feature experiment"s and "rollout experiment"s.  Specifically:
* "feature experiment" variations should contain "featureEnabled"
* "rollout experiment" variations should not contain "featureEnabled"

**Review by kellyroach-optimizely — CHANGES_REQUESTED** [review]
Added comments to the P.R. #214 requesting some improvement to the new unit test code comments (make these agree with the design doc).

File: OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m (hunk 5 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:423]: This was explicitly changed after @mikeng13 's [feedback](https://github.com/optimizely/ruby-sdk/pull/87/files/4e5f1b48c087bdff76eb0da0b48a357ec40b6ae4#diff-db360415c82dc1c3dd5a5c2872d0cd04R679).

Comment (mikeproeng37) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:423]: @abdurrafay is right. Rollout variations do contain the prop.

Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OptimizelyTest.m:423]: I accept Mike Ng's response.

**Comment by kellyroach-optimizely** [comment]
TEST:
* sh ./Scripts/build_all.sh PASSED
* sh ./Scripts/local_travis.sh PASSED
* 4x Analyze PASSED
* 8+1, 7+1 test suites PASSED
* 2 demo apps PASSED

**Review by kellyroach-optimizely — APPROVED** [review]

**Review by mikeproeng37 — APPROVED** [review]
lgtm

**Pull request merged**